### PR TITLE
Deprecate SpanContext::fromTraceparent() and allow passing custom sampling context data to startTransaction()

### PR DESCRIPTION
**Pull request by HazAT**
This is a PR to align the PHP SDKs to be equal on the implementation level as other SDKs.

The change to `fromTraceParent` is strictly speaking breaking and we also need to update Laravel SDK.
https://github.com/getsentry/sentry-laravel/pull/420

### CODE REVIEW

**Comment by ste93cry** [comment]
With the premise that I haven't looked at the code completely yet, why is it breaking? It looks to me that you just moved `SpanContext::fromTraceparent()` to `TransactionContext::fromSentryTrace()`, so we could handle this in a BC-compatible way by reporting a deprecation error in the parent method and overriding it in the subclass :thinking:

File: src/State/HubInterface.php
Comment (ste93cry) [inline src/State/HubInterface.php:145]: This is a huge breaking change, which is not acceptable without releasing a new major version. Rather than adding this argument for real, you can comment it and handle it with `func_get_arg` in the implementation class to avoid the break: we already did it in the past and it worked even though it was a bit messy

File: src/Tracing/SamplingContext.php
Comment (ste93cry) [inline src/Tracing/SamplingContext.php:23]: One blank line is more than sufficient 😄

File: src/Tracing/SamplingContext.php
Comment (ste93cry) [inline src/Tracing/SamplingContext.php:59]: Can you please add a little description of what this function does, e.g. `Sets additional optional context data.`?

File: src/Tracing/SamplingContext.php
Comment (ste93cry) [inline src/Tracing/SamplingContext.php:67]: Can you please add a little description of what this function does, e.g. `Gets additional optional context data.`?

File: src/Tracing/SpanContext.php
Comment (ste93cry) [inline src/Tracing/SpanContext.php:9]: This is a breaking change ⚠️ Instead of removing it, please add the following DocBlock to it

```php
/**
 * @deprecated since version 3.1, to be removed in 4.0
 */
```

File: src/Tracing/SpanContext.php
Comment (ste93cry) [inline src/Tracing/SpanContext.php:195]: This is a breaking change and cannot be done. Please leave the method as-is, mark is as deprecated and throw a `E_DEPRECATED` error following the standard of what has been done in `2.x`

File: src/State/Hub.php
Comment (ste93cry) [inline src/State/Hub.php:231]: Please add a blank line above this one to improve readability

**Review by ste93cry — CHANGES_REQUESTED** [review]

File: src/State/Hub.php
Comment (HazAT) [reply src/State/Hub.php:321]: We decided for `mixed` case in theory since it's a user defined function the user could return anything.

**Comment by HazAT** [comment]
Appveyor is failing because of rate limiting

File: src/Tracing/SpanContext.php
Comment (ste93cry) [inline src/Tracing/SpanContext.php:202]: Please add a blank line below to improve readability. Some day you will get this right 😄

File: src/Tracing/TransactionContext.php
Comment (ste93cry) [inline src/Tracing/TransactionContext.php:79]: Since the class is `final` we can now transform this to `new self()` and also add the typehint to the method rather than using the annotation

File: src/Tracing/TransactionContext.php
Comment (ste93cry) [inline src/Tracing/TransactionContext.php:9]: Please add a blank line below to improve readability.

File: src/State/Hub.php
Comment (ste93cry) [inline src/State/Hub.php:221]: Both add a new blank line above this one to improve readability and inline the `func_num_args()` function call

File: src/Tracing/SamplingContext.php
Comment (ste93cry) [inline src/Tracing/SamplingContext.php:60]: I think a better description would be `Sets additional data that will be provided as a second argument to {@link \Sentry\startTransaction()}.`

File: src/Tracing/SamplingContext.php
Comment (ste93cry) [inline src/Tracing/SamplingContext.php:70]: I think a better description would be `Gets the additional data that will be provided as a second argument to {@link \Sentry\startTransaction()}.`

File: tests/DsnTest.php
Comment (ste93cry) [inline tests/DsnTest.php:175]: As far as I know you cannot have two test cases with the same name/description. Please change this description to `invalid project ID (char instead of number)`

File: tests/DsnTest.php
Comment (ste93cry) [inline tests/DsnTest.php:180]: Please change this description to `invalid project ID (negative number)` to make it different from the one before

**Review by ste93cry — CHANGES_REQUESTED** [review]
Two things are still missing: the CHANGELOG entry and the `@expectedDeprecation` annotation on the failing test method (you can check the `2.x` branch to see how it works)

File: src/Tracing/SamplingContext.php
Comment (lobsterkatie) [inline src/Tracing/SamplingContext.php:65]: IDK how much this difference matters, but in the other SDKs, the context data provided in `customSamplingContext` is spread into `samplingContext`, whereas here it seems like it's being added as a single separate entry called `additionalContext`.

File: tests/Tracing/TransactionContextTest.php
Comment (lobsterkatie) [inline tests/Tracing/TransactionContextTest.php:44]:
```suggestion
    public function testFromTraceparent(string $header, ?SpanId $expectedSpanId, ?TraceId $expectedTraceId, ?bool $expectedParentSampled): void
    {
        $spanContext = TransactionContext::fromSentryTrace($header);

        if (null !== $expectedSpanId) {
            $this->assertEquals($expectedSpanId, $spanContext->getParentSpanId());
        }

        if (null !== $expectedTraceId) {
            $this->assertEquals($expectedTraceId, $spanContext->getTraceId());
        }

        $this->assertSame($expectedParentSampled, $spanContext->getParentSampled());
    }
```

File: src/Tracing/SamplingContext.php
Comment (HazAT) [reply src/Tracing/SamplingContext.php:65]: I would say this is ok to be not 100% the same, spreading this in PHP isn't something that feels natural in PHP.

Comment (ste93cry) [reply src/Tracing/SamplingContext.php:65]: My two cents: it's not really important whether we want to use the spreading operator or not, the key question is if this method should merge the new data with the existing one or not

Comment (lobsterkatie) [reply src/Tracing/SamplingContext.php:65]: That (merging) is how it works elsewhere. I feel like it's just a smidge nicer not to have to go digging any deeper than you need to for stuff, so I'd be in favor of the spread, but I don't feel strongly.

Comment (ste93cry) [reply src/Tracing/SamplingContext.php:65]: >spreading this in PHP isn't something that feels natural in PHP

Reading again @HazAT comment I think that what he meant is that in PHP a setter sets data, does not merge it, at least usually. I agree with this, but I don't have a strong opinion so I will let you two choose what to do. As far as I'm concerned, PR looks good and I'm waiting this discussion to reach an agreement before giving the final approval

Comment (lobsterkatie) [reply src/Tracing/SamplingContext.php:65]: >  in PHP a setter sets data, does not merge it

Ah, I see. Thanks for the explanation. Any reason we couldn't do away with this method entirely and just do

```PHP
$samplingContext = array_merge(SamplingContext::getDefault($context), $customSamplingContext));
```

in `startTransaction()`?

Comment (ste93cry) [reply src/Tracing/SamplingContext.php:65]: We cannot because `SamplingContext::getDefault()` return an object instance, so `array_merge` cannot work because you are trying to merge it with an array. Either we don't bother about the setter merging data (it's not a rule written in the stone that a setter must overwrite data rather than merging it) or we could rename the method to something else, e.g. `mergeAdditionalContext`

Comment (lobsterkatie) [reply src/Tracing/SamplingContext.php:65]: Okay. I poked at it a bit and I guess the real answer here would be to have `SamplingContext` just be a hashmap rather than a full-blown object instance, thereby allowing for arbitrary keys, but I assume that would be a totally breaking change, so... I think it's fine. Let's just leave it.

Thanks for the explanations - it's helpful to know a tiny bit about how PHP works.

**Review by lobsterkatie — APPROVED** [review]

**Review by ste93cry — APPROVED** [review]

**Pull request merged**
